app/src/debugger.lisp

```diff
@@ -0,0 +1,299 @@
+    (if (string= args "")
+        (funcall function)
+        (funcall function args))))
+
+(defun command-description (command)
+  (aref (nth-value 1 (cl-ppcre:scan-to-strings ".*\\n\\n(.*)" (command-documentation command))) 0))
+
+(define-debugger-command help (&optional command-name)
+  "Usage: help [COMMAND]
+
+Display help message."
+  (if command-name
+      (format t "~A~%" (command-documentation (get-command command-name)))
+      (loop :for (command-name . command) :in *commands* :do
+        (format t "~&~A~40T~A~%" command-name (command-description command)))))
+
+(define-debugger-command reset ()
+  "Usage: reset
+
+Reset the current QVM to the zero ket and set the program counter to zero."
+  (when *qvm*
+    (qvm::reset-quantum-state *qvm*)
+    (qvm::reset-classical-memory *qvm*)
+    (setf (qvm::pc *qvm*) 0)))
+
+(define-debugger-command list ()
+  "Usage: list
+
+Disassemble (list) the currently loaded program."
+  (unless *qvm*
+    (error "No program has been loaded."))
+  (loop :for instruction-number :from 0
+        :for instruction :across (qvm::program *qvm*) :do
+          (format t "~&~D~10T~/cl-quil:instruction-fmt/~%"
+                  instruction-number instruction)))
+
+(define-debugger-command run ()
+  "Usage: run
+
+Run the current program from the beginning."
+  (cond
+    ((and *qvm* (plusp (length (qvm::program *qvm*))))
+     (%reset)
+     (setf (qvm::pc *qvm*) 0)
+     (loop :for pc := (qvm::pc *qvm*)
+           :for breakpoint-p := (find pc *breakpoints*)
+           :until (or (null pc)
+                      (>= pc (qvm::loaded-program-length *qvm*))
+                      breakpoint-p)
+           :do (setf *qvm* (qvm:transition *qvm* (qvm::current-instruction *qvm*)))
+           :finally (when breakpoint-p
+                      (format t "Stopped at breakpoint before instruction:~%~
+                              ~D~10T~/cl-quil:instruction-fmt/~%"
+                              pc (qvm::current-instruction *qvm*)))
+                    (return *qvm*))
+     (when *display*
+       (%display-registers)))
+    (t
+     (error "No program has been loaded."))))
+
+(define-debugger-command display-registers ()
+  "Usage: display-registers
+
+Show the contents of the wavefunction after each executed step."
+  (format t "Amplitudes:~%")
+  (let ((nq (qvm:number-of-qubits *qvm*)))
+    (qvm:map-amplitudes
+     *qvm*
+     (let ((i 0))
+       (lambda (z)
+         (let ((p (* 100 (qvm:probability z))))
+           (when (> p *threshold*)
+             (format t
+                     "|~v,'0B>: ~/qvm-app::pprint-complex/, ~64TP=~5F~%"
+                     nq
+                     i
+                     z
+                     p)))
+         (incf i))))))
+
+(defun program-finished-p (qvm)
+  (or (null (qvm::pc qvm))
+      (>= (qvm::pc qvm) (qvm::loaded-program-length qvm))))
+
+(define-debugger-command step ()
+  "Usage: step
+
+Run the next instruction and stop."
+  (when *qvm*
```

Comment:
nitpick: some commands consider it an error if `*qvm*` is null (e.g. `%list`, `%run`), while others are happy to consider it a no-op.